opencog/reasoning/RuleEngine/rule-engine-src/README.md

```diff
@@ -0,0 +1,169 @@
+  3.The control mechanism, i.e. the policy for choosing which rules to apply in a given case,
+   must be pluggable. That is, when you invoke the rule engine, you get to choose which of the
+   available control policies will be used.   
+  4.There should be a standard way to express exclusivity between rules,and priority levels of rules --- but both of
+   these things should be relative to a given control policy (i.e rule X might have higher priority than rule Y with 
+   respect to policy A, but not policy B)   
+  5.The rule engine should be associated with some way of keeping track of which rules have been applied to which Atoms.
+    This information may be used by some control policies.   
+  6.Use pattern matcher for finding groundings and implement the callbacks if the need arises.
+    
+  further reading  
+  [http://wiki.opencog.org/w/Unified_Rule_Engin](http://wiki.opencog.org/w/Unified_Rule_Engin)  
+  [http://wiki.opencog.org/w/Control_policy](http://wiki.opencog.org/w/Control_policy)   
+  [http://wiki.opencog.org/w/Pattern_Matcher](http://wiki.opencog.org/w/Pattern_Matcher)
+   
+ **New PLN implementation overview**  
+  In general PLN rules and their associated formulas are all supposed to be ported to a scheme representation and are loaded at 
+  the beginning of the inference process(backward and forward chaining). most of PLN formulas have been ported
+in to a scheme file by contributors and can be found [here](https://github.com/opencog/opencog/tree/master/opencog/reasoning/RuleEngine/rules).
+
+  The high level algorithm for the new PLN forward and backward chaining is found [here](http://wiki.opencog.org/w/New_PLN_Chainer_Design).
+  
+**Algorithmic detail of the current implementation (Dec 2014)**
+  
+***Forward chaining***
+
+The implementation of the forward chainer is pretty straight forward.the call graph starts in
+the do_chain method.I have tried to use config files as a ways to declare what rules to use during the chaining process. the rules are read
+by load_fc_conf method. The current state of the code does the forward chaining algorithm stated in the wiki page with the exceptions of 
+ 
+Below is a kind of pseudo code of the do_bc function
+
+	do_chain(htarget)
+		hcurrent_target	
+		steps = 0
+		while (steps <= ITERATION_SIZE /*or !terminate*/) 
+			if steps == 0
+				if htarget == Handle::UNDEFINED
+					hcurrent_target = choose_target_from_atomspace(main_atom_space); //start FC on a random target
+				else
+					hcurrent_target = htarget;
+	    	else 
+				if (!target_list_.empty())
+					hcurrent_target = choose_target_from_list(target_list_)
+			choose_input(hcurrent_target); //add more premise via pattern matching of related atoms to hcurrent_target
+			choose_rule()
+			patternmatch(hcurrent_chosen_rule) // call the pattern matching with the chosen rule wrapped in a BindLink
+			steps++
```

Comment:
Oh, and another important case: what if the choosen rule yielded no results: i.e. it simply can't be applied to the current target?